### ISSUE TITLE
3rd User Study function expressable without out of bounds problem?

### ISSUE THREAD

**Issue by krisutofu** [iheartla/iheartla#119]
I was reading the paper and trying it myself. From the 3 user study functions, I had no problems with the first two but I couldn't get the last one to work.

The main reason is the lack of expression of constraints for the edges. The compiler generates code which generates invalid edges with negative indices. If it would at least allow natural numbers as constraints, then it could run in the dynamic language targets. For C++, it also needs an upper bound for edge indices and the guarantee of non-reflexive edges.

This is my try

```
A_ij = { 1 / |v_i - v_j| if (i,j) ∈ E
         0 otherwise

where
A ∈ ℝ^(n×n)
E ∈ { ℤ×ℤ }
v ∈ ℝ^(n)
```

It compiles but will fail with buffer overflow/out of bounds access. Instead of iterating over the index pairs of `A` it iterates over the edges of `E` which is very weird because `E` contains no constrained data.

This problem is severe. **This case will make I♥LA generate a potential security vulnerability!**
I would suggest: either disallow I♥LA programs where indices are retrieved from the if condtion or add some kind of constraint expression feature which logic operators.

**Comment by pressureless** [comment]
Thanks for your interest on I❤️LA. `A` is defined as a sparse matrix, that's why we try to iterate over the edges of `E`. I❤️LA doesn't provide run time checking, it relies on the user to prepare the correct parameters. By the way, I❤️LA uses 1-based indexing. If your parameter `E` has 0 using C++, then it will cause runtime error. Instead, you can either use `E ∈ { ℤ×ℤ } index` or making sure the parameters are larger than 0.

**Comment by krisutofu** [comment]
> ... Instead, you can either use `E ∈ { ℤ×ℤ } index` or making sure the parameters are larger than 0.

Crazy, I didn't know the `index` annotation. However, trying it will give no noticable benefit. The negative indices maybe are away now but C++ will still have a core dump due to out-of-bounds access. Python also has an out-of-bounds access. Of course, because it generates any random numbers that don't have to lie within the range of suitable indices.

Now, I don't know, how serious this problem actually is. Of course, the test data generation would not be integrated by users into their program. And there are goofy C++ programmers who choose performance over everything else like proper program functionality (safety) and reasonable security, i.e. using 'assert`s rather than actual error handling to prevent out-of-bounds access. For these goofy programmers, this behaviour is totally fine. For me, it wouldn't.

But when the assert is removed and `E` depends directly on user data, this would be a potential security vulnerability (at least an information leak). You may consider a better-security-option which throws/raises an exception (e.g.) or returns an error value whenever an input structure may be used for indexing.

**Comment by pressureless** [comment]
You're right. The current test data generation is not good enough, it doesn't make sure the data are valid. You can simply run multiple times when this happens. `asserts` are used for checking the types and dimensions are what users want. When the types go wrong, maybe it's better to directly let the user know. Users can also add error handling (e.g.: try/catch) when calling the generated code.
